README.md
# cablan_path_segment_fix
In June 2012, I discovered that the very common TYPO3 extension "Real URL" didn't properly store the path of pages in certain non-latin languages, even if TYPO3 already worked with them. I thus wrote a small extension to fix the missing paths which prevented the Japanese site of a client from being browsed. After testing, other languages were found to need the fix.


ChangeLog
2012-06-07 Martin-Pierre Frenette  <typo3@cablan.net>

	* initial code generated with kickstarter


pi1/class.tx_cablanpathsegmentfix_pi1.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2012 Martin-Pierre Frenette <typo3@cablan.net>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/
require_once(PATH_tslib.'class.tslib_pibase.php');


/**
 * Plugin 'real url pathsegment fix' for the 'cablan_path_segment_fix' extension.
 *
 * This extension repairs the path segements for many languages which TYPO3 is unable
 * to properly process.
 *
 * This was discovered only in 2012 when a client began using realurl with non latin language!
 *
 *
 * @author	Martin-Pierre Frenette <typo3@cablan.net>
 * @package	TYPO3
 * @subpackage	tx_cablanpathsegmentfix
 */
class tx_cablanpathsegmentfix_pi1 extends tslib_pibase {
	var $prefixId      = 'tx_cablanpathsegmentfix_pi1';		// Same as class name
	var $scriptRelPath = 'pi1/class.tx_cablanpathsegmentfix_pi1.php';	// Path to this script relative to the extension dir.
	var $extKey        = 'cablan_path_segment_fix';	// The extension key.
	var $pi_checkCHash = true;
	


	function getRecord($table, $uid, $enableFields = 1){
        $where = ' uid=' . intval($uid);
        if ($enableFields) {
            $where .= $this->cObj->enableFields($table);
        }

        $result = $GLOBALS['TYPO3_DB']->exec_SELECTquery('*', $table, $where);
        return @$GLOBALS['TYPO3_DB']->sql_fetch_assoc($result);
    }
		
	/**
	 * The main method of the PlugIn. It scan the path segment for paths
	 * which are empty because they failed to generate!
	 *
	 * @param	string		$content: The PlugIn content
	 * @param	array		$conf: The PlugIn configuration
	 * @return	The content that is displayed on the website
	 */
	function main($content, $conf)	{

		$res = $GLOBALS['TYPO3_DB']->exec_SELECTquery('sys_language.uid,sys_language.flag, static_languages.lg_iso_2', 'sys_language,static_languages', 'static_lang_isocode=static_languages.uid');

		$languages_to_fix = $this->GetNonLatinLanguageList();

		while ( $row=  $GLOBALS['TYPO3_DB']->sql_fetch_assoc($res)){

			if (in_array($row['lg_iso_2'], $languages_to_fix) ){
				$langs[] =$row['uid'];
			}
		}

		$res = $GLOBALS['TYPO3_DB']->exec_SELECTquery('*',  'pages_language_overlay', 
			'length(tx_realurl_pathsegment) =0 AND sys_language_uid IN ('.implode(',',$langs). ')');

		while ( $row=  $GLOBALS['TYPO3_DB']->sql_fetch_assoc($res)){
			$page = $this->getRecord('pages',$row['pid'] );
			$values = array();
			if ( strlen($page['tx_realurl_pathsegment']) > 0){
				$values['tx_realurl_pathsegment'] = $page['tx_realurl_pathsegment'];
			}
			else{
				$values['tx_realurl_pathsegment'] = str_replace(' ', '-', $page['title']);	
			}

			$GLOBALS['TYPO3_DB']->exec_UPDATEquery(  'pages_language_overlay', 
				'uid='. $row['uid'], $values);			
		}
	}

	function GetNonLatinLanguageList(){

		return array( 
			'ZH', // chinese
			'AR', // arabic,
			'MY', // burmese
			'HE', // Hebrew
			'HI', // Hindi
			'JA', // Japanese
			'KO', // Korean
			'FA', // Persian
			'PA', // Punjabi
			'RU', // Russian
			'TR', // turkish
			'VI', // vietnamese

			);
	}
}



if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/cablan_path_segment_fix/pi1/class.tx_cablanpathsegmentfix_pi1.php'])	{
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/cablan_path_segment_fix/pi1/class.tx_cablanpathsegmentfix_pi1.php']);
}

?>

ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "cablan_path_segment_fix".
#
# Auto generated 07-06-2012 09:28
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Cablan path segment fix for non latin languages',
	'description' => 'This extension automatically set the default language page title as the Real url path segment for non latin languages.

This prevents Japanese, Chinese, Vietnamese or other similar languages from preventing',
	'category' => '',
	'author' => 'Martin-Pierre Frenette',
	'author_email' => 'typo3@cablan.net',
	'shy' => '',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'module' => '',
	'state' => 'alpha',
	'internal' => '',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearCacheOnLoad' => 0,
	'lockType' => '',
	'author_company' => '',
	'version' => '0.0.0',
	'constraints' => array(
		'depends' => array(
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:7:{s:9:"ChangeLog";s:4:"6825";s:10:"README.txt";s:4:"ee2d";s:12:"ext_icon.gif";s:4:"1bdc";s:17:"ext_localconf.php";s:4:"3553";s:19:"doc/wizard_form.dat";s:4:"9c24";s:20:"doc/wizard_form.html";s:4:"f8bf";s:41:"pi1/class.tx_cablanpathsegmentfix_pi1.php";s:4:"f56b";}',
);

?>